rawterm.h

```diff
@@ -39,572 +39,528 @@
-std::string underline(const std::string& s) { return "\x1b[4m" + s + "\x1b[24m"; }
-std::string blink(const std::string& s) { return "\x1b[5m" + s + "\x1b[25m"; }
-std::string inverse(const std::string& s) { return "\x1b[7m" + s + "\x1b[27m"; }
-std::string hidden(const std::string& s) { return "\x1b[8m" + s + "\x1b[28m"; }
-std::string strikethrough(const std::string& s) { return "\x1b[9m" + s + "\x1b[29m"; }
-
-// Cursor positioning
-
-#endif // RAWTERM_IMPLEMENTATION
+namespace rawterm {
+	namespace detail {
+		inline termios orig;
+
+		inline void die(const char* s) {
+			std::perror(s);
+			std::exit(1);
+		}
+	} // namespace detail
+
+	enum struct Mod {
+		Alt_L,
+		Arrow,
+		Backspace,
+		Control,
+		Delete,
+		Escape,
+		Function,
+		Shift,
+		Space,
+		Unknown
+	};
+
+	struct Key {
+		char code;
+		std::vector<rawterm::Mod> mod;
+		std::string raw;
+	};
+
+	struct Pos {
+		int line;
+		int col;
+	};
+
+	// Enter/leave alternate screen
+	// https://stackoverflow.com/a/12920036
+	// Will need to find another solution for windows (#ifdef WIN32)
+
+	// This doesn't need explicitly calling because of the atexit() call
+	inline void disable_raw_mode() {
+		if (tcsetattr(STDIN_FILENO, TCSAFLUSH, &rawterm::detail::orig) == -1) {
+			rawterm::detail::die("tcsetattr");
+		}
+	}
+
+	inline void enable_raw_mode() {
+		if (tcgetattr(STDIN_FILENO, &rawterm::detail::orig) == -1) {
+			rawterm::detail::die("tcgetattr");
+		}
+		std::atexit(rawterm::disable_raw_mode);
+
+		termios raw = rawterm::detail::orig;
+		cfmakeraw(&raw);
+
+		if (tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw) == -1) {
+			rawterm::detail::die("tcsetattr");
+		}
+	}
+
+	inline void enter_alt_screen() {
+		std::cout << "\x1B 7\x1B[?47h\x1B[H";
+	}
+
+	inline void exit_alt_screen() {
+		std::cout << "\x1B[2J\x1B[?47l\x1B 8";
+	}
+
+	// Read user input
+	inline rawterm::Key process_keypress() {
+		static const std::unordered_set<char> asciiLetters {
+			'\x31', '\x32', '\x33', '\x34', '\x35', '\x36', '\x37', '\x38',
+			'\x39', '\x41', '\x42', '\x43', '\x44', '\x45', '\x46', '\x47',
+			'\x48', '\x49', '\x4A', '\x4B', '\x4C', '\x4D', '\x4E', '\x4F',
+			'\x50', '\x51', '\x52', '\x53', '\x54', '\x55', '\x56', '\x57',
+			'\x58', '\x59', '\x5A', '\x61', '\x62', '\x63', '\x66', '\x67',
+			'\x66', '\x67', '\x68', '\x69', '\x6A', '\x6B', '\x6C', '\x6D',
+			'\x6E', '\x6F', '\x70', '\x71', '\x72', '\x73', '\x76', '\x77',
+			'\x76', '\x77', '\x78', '\x79', '\x7A'
+		};
+
+		std::string characters;
+		const int ret = read(STDIN_FILENO, characters.data(), 5); // max length of esc sequence
+		if (ret < 0) {
+			std::cerr
+				<< "ERROR: something went wrong during reading user input: "
+				<< std::strerror(errno) << std::endl;
+			return {' ', {rawterm::Mod::Unknown}, ""};
+		}
+
+		std::stringstream ss;
+		ss << std::uppercase << std::hex << std::setw(2) << std::setfill('0');
+		for (char c : characters) {
+			ss << "\\x" << static_cast<int>(c);
+		}
+		std::string raw = ss.str();
```

Comment:
I'm not sure if lines 137-142 are working as intended, the `raw` variable is empty when printed out at all. Quicky print debugging this and it seems like `characters.data()` on line 129 might not be doing what it's meant to? Printing out `characters` just above the stringstream line shows that variable is empty

Reply:
I don't know if this would work exactly, but could we replace line 129 with a `std::cin`?

Reply:
hmm

Reply:
I couldn't get `std::cin` to work like [here](https://github.com/Eczbek/xieite/blob/main/include/xieite/streams/standard_handle.hpp#L236-L242)

Reply:
Even after the changes, I'm still running into some issues around this section of code -- pressing `asdf` should only print out the first hex in each raw instead of all of it:
```
Key{ code: a, mods: [], raw: \x61\xB2\x5A\xF7\xA\x1}
Key{ code: s, mods: [], raw: \x73\xB2\x5A\xF7\xA\x1}
Key{ code: d, mods: [], raw: \x64\xB2\x5A\xF7\xA\x1}
Key{ code: f, mods: [], raw: \x66\xB2\x5A\xF7\xA\x1}
```

I think it might be worth reverting these changes for now and coming back to these in another PR, so I can merge in the other changes and move ahead with this. I think we just need to work out what the proper modern version of [this chunk of code](https://github.com/Ttibsi/rawterm/blob/main/rawterm.h#L231-L249) should be -- I think you're on the right lines here, I'd be surprised if it couldn't use more modern-isms like `std::string` somehow though -- let me know what you think here?

Reply:
Alternatively -- so this code came straight from [here](https://github.com/snaptoken/kilo-src/blob/error-handling/kilo.c#L40), which is a C text editor guide. I've found a version online rewritten in c++ [here](https://github.com/r-darwish/kilopp/blob/25f6bb9ae8780a93f77d5c4c8b7ad407370b6db8/kilo.cpp#L340-L344) -- what if we take the same  approach here, instead of using a `std::string`, we use an `std::array` to replace the char array and update `read()` accordingly? Thoughts?